Customizable serialization of PostgresType

yrashk opened the pull request:
Problem: CBOR is not the most universal answer to everything

However, all `Serialize/Deserialize` PostgresTypes are forced to use with the blanket implementation of `IntoDatum`/`FromDatum`

Solution: extract a `Serializer` and `Deserializer` traits to abstract this away

By default, it uses CBOR and encodes it into a Varlena, replicating existing behavior.

However, this gives a good amount of flexibility to end-user to determine their encoding needs.

I've also removed JSON-encoding/decoding functions from varlena module, as they don't seem to be used anywhere.

---

A message to pgx users: my ability to actively contribute to pgx largely depends on your support. Please consider [sponsoring my work](https://github.com/sponsors/yrashk)

yrashk committed:
Problem: CBOR is not the most universal answer to everything

However, all `Serialize/Deserialize` PostgresTypes are forced to use it
with the blanket implementation of `IntoDatum`/`FromDatum`

Solution: extract a `Serializer` trait to abstract this away

By default, it uses CBOR and encodes it into a Varlena, replicating
existing behavior.

However, this gives a good amount of flexibility to end-user to
determine their encoding needs.

I've also removed JSON-encoding/decoding functions from varlena module,
as they don't seem to be used anywhere.

yrashk committed:
Problem: missing documentation for pgx::ddatum::Serializer

Makes it unclear whether it's advisable to use it externally.

Solution: document it

yrashk committed:
Problem: pgx::datum::Serializer trait scope

It takes care of both serialization and deserialization. However, some
types can potentially be only deserialized, or serialized. Besides, the
naming of the trait itself showed that it is not a perfect match.

Solution: split it into Serializer and Deserializer, respectively

Smittyvb commented:
This is useful for [the pgx extension I work on](https://github.com/timescale/timescaledb-toolkit), since we use a non-CBOR serialization for our aggregate states (Bincode) and it would be great if we were able to use `pg_aggregate` for them.

eeeebbbbrrrr commented:
hey @yrashk!  this fell off my radar somehow.  I think it looks good in general.  Would you mind rebasing so we can see it without any conflicts?  I'd like to merge if we can

yrashk commented:
Absolutely! Will look into this shortly.

yrashk commented:
Here's my first take on rebasing. It still requires a bit more of a review as things have changed and I may have missed something. The test passes.

yrashk committed:
Problem: custom_serializer test is not checking serialization

For all we know, it may still be using CBOR.

Solution: ensure we're getting our custom serialization

yrashk commented:
@eeeebbbbrrrr did you have a chance to look at this updated PR yet?

eeeebbbbrrrr commented:
I have not. :( I’ll tackle it next week. Thanks for the ping.
